[magentatv] Initial contribution (aka EntertainTV) (#7864)
openhab/openhab-addons: bundles/org.openhab.binding.magentatv/README.md

Standby provides the best results, because the binding could wake-up the receiver (Power On/Off). Suspend/Resume would require a Wake-on-LAN, which could be done, but is currently not implemented. 
Shutdown turns the receiver off, which requires a manual PowerOn.

There is no way to detect the "display status" of the receiver. 
The binding detects Power-Off with the MR401B/MR201 by listening to UPnP events, but can't verify the status when started.
You need to take care on the current status if you power on/off the receiver from scenes.
Check the current status before sending the POWER button, because POWER is a toggle, not ON or OFF (see sample rules).

### Discovery

The auto discovery starts when the binding is loaded.
UPnP will be used to discover receivers on the local network.
Based on UPnP all necessary parameters are auto-discovered.
The receiver needs to be powered on to get discovered.

### Supported Things

|Thing    |Type                                                           |
|---------|-------------------------------------------------------------- |
|receiver |A MagentaTV Receiver, the binding supports multiple receivers. |

> # Device setup

Reviewer: ```suggestion
# Device Setup
```

Author: changed

> Once the thing will be added from the Inbox in PaperUI you'll need your T-Online credentials to query the userID.

Reviewer: Please do not mention Paper UI anywhere as this soon won't exist anymore.

Author: removed

> Once the thing will be added from the Inbox in PaperUI you'll need your T-Online credentials to query the userID.

Reviewer: It is not clear what "userID" refers to. The thing requires an udn as a parameter, is it this what is meant?

Author: I replaced it with UID to avoid confusion

In fact we binding needs to login to the portal and run the OAuth process. As results of an successful login you'll receive a so called userId, which is required to do the local pairing.

You could obtain the ID with Wireshark and set it up manually in the thing config, but of course most users have the receiver connected to the Internet anyways.

> Open the thing configuration and enter the credentials:
> 
> |Field            |Description                                      |
> |---------------- |------------------------------------------------ |
> |Account Name     |Your T-Online user id, e.g. test7017@t-online.de |
> |Account Password |The password for your Telekom account.           |

Reviewer: ```suggestion
```
The thing configuration parameters are already listed below.

Author: removed

> For security reasons the credentials will be automatically deleted from the thing configuration (replaced with '***' in the thing config) after the initial authentication process.

Reviewer: There might be a much more elegant solution for that: Create a console command that does the login and returns the parameter that is required for the thing configuration (what exactly is it?).
Check the Tesla binding (https://www.openhab.org/addons/bindings/tesla/#bridge-configuration), which offers this way (additionally to putting the credentials directly in the Thing configuration).

Author: who wants to use the console?

Reviewer: Who wants to enter his password in a web UI, which is not related to the service the password belongs to? That's against any security best practices.

Author: @kai why does it not belong to the service? In fact that's a usual OAuth flow, which provides a token that is required for pairing. If credentials should never stored in textual thing configs we need to change a lot of bindings. What do you request?

Author: refactored, see comment below

The openHAB instance needs access to the Internet to perform that operation, which can be disabled afterwards.

> One the userID has been obtained from the Telekom portal the binding initiates the pairing with the receiver.
> This is an automated process.
> The thing changes to ONLINE state once the pairing result is received.
> Otherwise open PaperUI:Confuguration:Things and check for an error message.
> Using [Show Properties] you see more details and could verify if the discovery and pairing were completed successful.

Reviewer: ```suggestion
```

Author: refactored, see comment below

> The binding uses the network settings in openHAB system configuration to determine the local ip address. The device can't be discovered when the openHAB system and receiver are not on the same network (ip/netmask).

Reviewer: ```suggestion
The binding uses the network settings in openHAB system configuration to determine the local IP address. The device cannot be discovered if the openHAB system and receiver are not on the same network (ip/netmask).
```

Author: changed

## Thing Configuration

|Parameter       |Description                                                                                                     |
|----------------|----------------------------------------------------------------------------------------------------------------|
|udn             |UPnP Unique Device Name - a hex ID, which includes the 12 digit MAC address at the end (parsed by the binding)  |

> |modelId         |Type of Media Receiver: DMS_TPB for MR400 and MR200 ; MR401B for MR401B and MR201                               |

Reviewer: Why is this parameter required? If it makes a difference, you should model it as different thing types, that's exactly what they are meant for.

Author: modelId is removed, was a left-over from initial implementation

# MagentaTV Program Information
String MagentaTV_ProgTitle   "Program Title" {channel="magentatv:receiver:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx:program#title"}
String MagentaTV_ProgDescr   "Description"   {channel="magentatv:receiver:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx:program#text"}
String MagentaTV_ProgStart   "Start Time"    {channel="magentatv:receiver:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx:program#tart"}
String MagentaTV_ProgDur     "Duration"      {channel="magentatv:receiver:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx:program#duration"}
String MagentaTV_ProgPos     "Position"      {channel="magentatv:receiver:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx:program#position"}

# MagentaTV Play Status
Number MagentaTV_ChCode    "Channel Code"    {channel="magentatv:receiver:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx:status#channelCode"}
String MagentaTV_PlayMode  "Play Mode"       {channel="magentatv:receiver:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx:status#playMode"}
String MagentaTV_RunStatus "Run Status"      {channel="magentatv:receiver:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx:status#runStatus"}
```

### sitemap

```

> please contribute an example

Reviewer: yes, please do so.

Author: done

```

### magentatv.rules

Due to the fact the POWER is a toggle button and the binding can't detect the current status, which could lead into the situation that you want to power on the receiver as part of a scene, but due to the fact that it is already ON you switch it off.
We spend some time to fiddle out a better handling and find a way to detect a network message when the receiver gets powered off (MR4xx only).
In this case MagentaTV_Power is switch to OFF.

> This said you could use the following

Reviewer: Sorry, I do not understand this...

Author: changed, please re-visit

# MagentaTV Binding

> This Binding allows controlling the Deutsche Telekom Magenta TV Media Receiver series MR4xx and MR2xx (Telekom NGTV / Huawei Envision platform). 

Reviewer: ```suggestion
This binding allows controlling the Deutsche Telekom Magenta TV Media Receiver series MR4xx and MR2xx (Telekom NGTV / Huawei Envision platform). 
```

Author: changed

> The Binding does NOT support MR3xx/1xx (old Entertain system based on Microsoft technology)!

Reviewer: ```suggestion
The binding does NOT support MR3xx/1xx (old Entertain system based on Microsoft technology)!
```

Author: changed

Media Receivers are automatically discovered.

> You could send keys as you press them on the remote and the binding receives program information when the channel is switched (no

Reviewer: ```suggestion
You can send keys as you press them on the remote and the binding receives program information when the channel is switched (no
```

Author: changed

* Deutsche Telekom Media Receiver MR200  - supported with minor restrictions (POWER key)
* Deutsche Telekom Media Receiver MR601  - should be supported (not verified)
* Deutsche Telekom Media Receiver MR3xx  - NOT supported (different platform)
* Deutsche Telekom Media Receiver MR1xx  - NOT supported (different platform)

## Auto Discovery

UPnP will be used to discover receivers on the local network and discover the nessesary parameters.
The receiver needs to be powered on to get discovered.

Once the receiver is discovered it can be added from the Inbox.
Make sure to set the parameter `userId` in the Thing configuration after adding the new thing, see section Thing Configuration.

Note:
The binding uses the network settings in openHAB's system configuration to determine the local IP address.
The device can't be discovered if the openHAB system and receiver are not on the same network (IP/Netmask).
In this case you need to add the Thing manually or use textual configuration (.things).

If you are running openHAB in a Docker container you need to make sure that UPnP discovery is available and network interfaces 

> ## Generating the UID (`userId`)

Reviewer: It is not being "generated", but rather retrieved.

Wouldn't it make sense to always just use one name for it? You are now mixing UID and userId everywhere. I think it would be easier for users, if you'd simply talk about the `userId` parameter throughout the documentation.

This section must not be in the "auto discovery" section, but rather in the "Thing configuration" below.

Author: done

The binding requires a so called UID (`userId`) to pair with the receiver, which is generated by logging in with your Telekom credentials (Login Name + Password for the "Telekom Kundencenter").
openHAB needs to access the Internet to perform that operation, which can be disabled afterwards.

> The binding initiates the pairing with the receiver once the userId has been obtained, this is an automated process.

Reviewer: Is there really some "pairing" process happening here? If so, when? Everytime the thing handler is started? Or is it simply an "establishing connection" process?

Author: it creates some kind of logical connection between the receiver and the applcation. The MR sends the event updates to the applcation. This get renewed every time the binding starts and renewed periodically. It's a 3-stage process.

The Thing changes to ONLINE once the pairing result is received, otherwise check the Thing status for an error message.

There are different ways to setup the userId:

1. Use the openHAB console
Run the following command on the console and provide your Tesla account credentials (the same that you use in the official Tesla app):

```
openhab> smarthome:magentatv login
Username (email): mail@example.com
Password: topsecret

Attempting login...
Login successful, returned UID (userId) is 1903AAAAAAAAC7E9718BBBCBCBCBCBC
Edit thing configuration and copy this value to the field UID
```

On successful login the console will show the userId value. Copy&amp;Paste this value to the Thing configuration (parameter UID / `userId`) of the receiver.

> If you have multiple receivers under the same MagentaTV subscription you could re-use this value for the other receiver Things. 

Reviewer: ```suggestion
If you have multiple receivers under the same MagentaTV subscription you can use this value for all these receivers. 
```

Author: changed

2. Provide your credentials in the UI
If you do not want to use the openHAB console, you can also setup the credentials in the Thing configuration 

- Account Name (`accountName`) is your Login Name for the Telekom Kundencenter (registered email address) 
- Account Password (`accountPassword`) is the corresponding password.

> The binding will do an automated login, receives the UID (`userId`) and stores this in the Thing configuration.

Reviewer: ```suggestion
The binding uses these credentials to login to your account, receives the `userId` parameter and sets it in the Thing configuration.
```

Author: changed

> For security reasons the credentials will be automatically deleted from the thing configuration (replaced with '***') after the initial process.

Reviewer: ```suggestion
For security reasons the credentials are automatically deleted from the thing configuration (replaced with '***') after the initial process.
```
Err, why replaced with '***'? The two parameters should be set to null instead.

Author: done

* Deutsche Telekom Media Receiver MR200  - supported with minor restrictions (POWER key)
* Deutsche Telekom Media Receiver MR601  - should be supported (not verified)
* Deutsche Telekom Media Receiver MR3xx  - NOT supported (different platform)
* Deutsche Telekom Media Receiver MR1xx  - NOT supported (different platform)

## Auto Discovery

UPnP will be used to discover receivers on the local network and discover the nessesary parameters.
The receiver needs to be powered on to get discovered.

Once the receiver is discovered it can be added from the Inbox.
Make sure to set the parameter `userId` in the Thing configuration after adding the new thing, see section Thing Configuration.

Note:
The binding uses the network settings in openHAB's system configuration to determine the local IP address.
The device can't be discovered if the openHAB system and receiver are not on the same network (IP/Netmask).
In this case you need to add the Thing manually or use textual configuration (.things).

If you are running openHAB in a Docker container you need to make sure that UPnP discovery is available and network interfaces 

## Receiver Standby Mode

> The Media receiver has 3 different standby modes, which could be selected in the receiver's settings.

Reviewer: ```suggestion
The Media receiver has 3 different standby modes, which can be selected in the receiver's settings.
```

Author: changed

|Mode          |Description                                                                             |
|--------------|----------------------------------------------------------------------------------------|

> |Standby       |Full standby - the receiver active all the time, sleeps when off and wakes up instantly.|

Reviewer: ```suggestion
|Standby       |Full standby - the receiver is active all the time, even while sleeping, so it can wake up instantly.|
```

Author: changed

> |Suspend/Resume|The receiver goes to sleep mode, but could be awaked by a Wake-on-LAN packet            |

Reviewer: ```suggestion
|Suspend/Resume|The receiver goes to sleep mode, but can be awakened by a Wake-on-LAN packet            |
```

Author: changed

> |Shutdown      |Powering off will shutdown the receiver, can be awaked only with the power button       |

Reviewer: ```suggestion
|Shutdown      |Powering off shuts down the receiver, so that it can be awakened only with the power button       |
```

Author: done

> `Standby` provides the best results, because the binding could wake-up the receiver (Power On/Off).

Reviewer: ```suggestion
`Standby` provides the best results, because the binding can wake up the receiver (Power On/Off).
```

Author: changed

> `Suspend/Resume` would require a Wake-on-LAN, which could be done, but this is currently not implemented. 

Reviewer: ```suggestion
`Suspend/Resume` requires a Wake-on-LAN packet, which can take longer.
```

Author: changed

| MENU     | Menu                                           |
| INFO     | Display information                            |
| FAV      | Display favorites                              |
| VOLUP    | Volume up                                      |
| VOLDOWN  | Volume down                                    |
| MUTE     | Mute speakers                                  |
| CHUP     | Channel up                                     |
| CHDOWN   | Channel down                                   |
| PLAY     | Play                                           | 
| PAUSE    | Play                                           | 
| STOP     | Stop playing                                   |
| RECORD   | Start recording                                |
| REWIND   | Rewind                                         |
| FORWARD  | Forward                                        |
| LASTCH   | Last channel                                   |
| NEXTCH   | Next channel                                   |
| LASTCHAP | Last chapter                                   |
| NEXTCHAP | Next chapter                                   |
| PAIR     | Re-pair with the receiver                      |

> In addition you could send any key code in the 0xHHHH format., refer to

Author: changed

String MagentaTV_ProgStart   "Start Time"    {channel="magentatv:receiver:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx:program#tart"}
String MagentaTV_ProgDur     "Duration"      {channel="magentatv:receiver:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx:program#duration"}
String MagentaTV_ProgPos     "Position"      {channel="magentatv:receiver:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx:program#position"}

# MagentaTV Play Status
Number MagentaTV_ChCode    "Channel Code"    {channel="magentatv:receiver:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx:status#channelCode"}
String MagentaTV_PlayMode  "Play Mode"       {channel="magentatv:receiver:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx:status#playMode"}
String MagentaTV_RunStatus "Run Status"      {channel="magentatv:receiver:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx:status#runStatus"}
```

### sitemap

```
please contribute an example
```

### magentatv.rules

Due to the fact the POWER is a toggle button and the binding cannot detect the current status, which could lead into the situation that you want to power on the receiver as part of a scene, but due to the fact that it is already ON you switch it off.

> Beginning with models 401/201 and new the binding is able to detect the Power-OFF condition, which can be used in a rule to improve this situation

Author: The binding is not able to read the status. The only improvement I found was to catch the UPnP event on power-off + POWER is a toggle button, you can't switch ON or OFF you can only toggle the current status. In combination with the fact that the status is unknown you depend on the current status. I also had a version, which tries to keep the status in the binding, but practically this doesn't work out well.

> Beginning with models 401/201 and new the binding is able to detect the Power-OFF condition, which can be used in a rule to improve this situation

Reviewer: That still does not explain what the rule is meant to do other than what could be implemented in the binding already.

> Beginning with models 401/201 and new the binding is able to detect the Power-OFF condition, which can be used in a rule to improve this situation

Author: If the binding is not transparent and the status is not in sync you are never able to send the POWER key, e.g.
- OH starts, receiver is off
- OH sends an POWER, receiver goes ON
- now you press the power button on the remote = receiver OFF vs. binding state ON
- you run a scene, which expects the receiver to be ON
- the binding doesn't send ON, because it's status is still ON

With the MR401 the beding detects the manual POWER OFF, but not with the MR400
In the current implementation you could bring both in sync just by sending POWER manually, which is send to the receiver, because the binding is transparent.

> Beginning with models 401/201 and new the binding is able to detect the Power-OFF condition, which can be used in a rule to improve this situation

Reviewer: So what you are saying is that although you can identify the power state on an MR401, this state is still unreliable, so that the user needs a way to bring the real receiver state and the item state back in sync (and this is only possible if the logic is not in the binding, but only in external rules)?

> Beginning with models 401/201 and new the binding is able to detect the Power-OFF condition, which can be used in a rule to improve this situation

Author: ok, we have

MR 400/MR200
- does not send SSDN messages when shutting down
- responds to ping when network interface is active, but takes about 45s on start-up
- no way to detect display status or state of the power button

MR 401/201
- sends SSDN message on shutdown, but nothing on startup
- does not respond to ping - so, this can't be used to detect power status
- no way to detect display status or state of the power button
- answers to http requests event when in state OFF (not in sleep mode) -> can't use dummy request to see if the device is awake
- sometimes does an UPnP advertisement (which is intercepted by the binding)
- sends program events when a channel is switched, but not initially when the MR is turned ON

+ 
- POWER is a toggle button
- if you want to start your scene while the receiver is ON you'll switch it OFF

I think it's a bad idea to build a logic into the binding, which relies on SSDN shutdown message, but is not able to detect that the receiver is ON, we need to differ between
- MR is in deep sleep / not connected to power
- MR is in standby mode, responds to network requests, but doesn't report the POWER status
- MR is active and playing a TV channel
- MR is active and playing streaming content

The SSDN message helps to get the POWER status in sync, but I didn't find a way to see that the MR is not only responding to network request, but is aware and playing content. In my UI I have a button sending POWER and one starting a scene (switching TV and MR on, selecting HDMI etc.). The scene uses the rule and usually I need only that button. The rule checks (the binding's) POWER state and sends ON if the MR is currently off. If for any reason the OH status and MR status get out of sync  I have the "raw" POWER button to fix that.

I spend several hours to find any indicator, which helps to overcome this issue, but without success. Maybe you have the magic idea. Any by the way: Nobody is obligated to use the external rule, but an incomplete logic in the binding could screw up the user experience with no way to change it.

```

On successful login the console will show the User ID value. Copy&amp;Paste this value to the Thing configuration (parameter `userId`) of the receiver.
If you have multiple receivers under the same MagentaTV subscription you can use this value for all of them. 

2. Provide your credentials in the UI

If you do not want to use the openHAB console, you can also setup the credentials in the Thing configuration 

- Account Name (`accountName`) is your Login Name for the Telekom Kundencenter (registered email address) 
- Account Password (`accountPassword`) is the corresponding password.

The binding uses these credentials to login to your account, retrieves the `userId` parameter and sets it in the Thing configuration.
For security reasons the credentials are automatically deleted from the thing configuration after the initial process.

## Channels

|Group   |Channel        |Item-Type|Description                                                               |
|--------|---------------|---------|--------------------------------------------------------------------------|
|control |power          |Switch   |Switching the channel simulates pressing the power button (same as sending 

> "POWER" to the key channel). The receiver doesn't offer ON and OFF, but just toggles the power state.

Author: changed
